### DISCUSSION TITLE
Cancel wire:stream request

### DISCUSSION THREAD

**Discussion by andrejguran** [livewire/livewire#7882] [1. Help]
In [documentation](https://livewire.laravel.com/docs/wire-stream#streaming-chat-bot-responses) there is a handy use-case for Chat bot implementation using `wire:stream`. There is a common patter with chatbots to be able to stop generating response. When trying to implement this I got stuck on livewire blocking all requests until the stream request is finished. Is there a way to abort a streaming request in livewire?

**Comment by kaburagisec** [comment]
send the full code here, we will help according your code

**Reply by andrejguran** [reply]
The question is how in the [documentation](https://livewire.laravel.com/docs/wire-stream#streaming-chat-bot-responses)'s `ChatBot` class example implement a stop button that will abort the streaming request

**Comment by giltotherescue** [comment]
I am also looking for a solution to this problem. It causes [an issue](https://github.com/livewire/livewire/discussions/7009) with wire:navigate if the user navigates to a different livewire view, then back to the original view, while the stream is still processing.

**Comment by SmoKE585** [comment]
me to :( anyone have solution?

**Comment by svemir** [comment]
I was so happy to find my exact question, and yet disappointed to also find my current answer as well, which is not much...

**Comment by giltotherescue** [comment]
Anyone found a solution yet?

**Comment by kaburagisec** [comment]
My solution is to make a request from stopButton with x-on:click which triggers the API to change the chat data and in the stream there is a check whether the stream is true or false.

Example:

```
// blade.php:
<button x-on:click="fetch('/api/stop-stream', {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json',
                            'X-CSRF-TOKEN': document.querySelector('meta[name=csrf_token]').content
                        },
                        body: JSON.stringify({ stream: false, uuid: uuid }),
                        credentials: 'same-origin'
                    })"> </button>
```

```
// api.php
use App\Livewire\ChatBot;
    
Route::post("/stop-stream", [ChatBot::class, "stop_prompt"]);
```

```
//Chatbot.php
public function stop_prompt(Request $request)
{
    if(!$request->uuid){
        return;
    }
        
    $chat_index = count($this->chat) - 1;
    $this_chat = Chat::where('uuid', $request->uuid)->first();
    if($this_chat){
        $this_chat->stream = false;
        $this_chat->save();
        return 'ok';
    } else{
        return 'not ok';
    }
}

public function send_prompt_ollama($prompt, $uuid)
{
    $response = Http::asJson()->withOptions([
            'stream' => true
        ])->post("http://localhost:11434/api/chat", [
            "model" => $this->llm_model,
            "messages" => $formattedMessages
        ]);

    $body = $response->getBody();
        
    $buffer = '';
        
    while (!$body->eof()) {
        $buffer .= $body->read(512);
    
        while (($pos = strpos($buffer, "\n")) !== false) {
            if(!Chat::where('uuid', $this->uuid)->value('stream')) {
                break 2;
            }
            $line = trim(substr($buffer, 0, $pos));
            $buffer = substr($buffer, $pos + 1);
                
            if ($line === '') continue;
    
            $data = json_decode($line, true);
            if (isset($data['message']['content'])) {
                $chunk = $data['message']['content'];
                $html_response = $this->customParser1(Markdom::toHtml($chunk));
                $this->stream(to: "answer-$chat_index", content: $html_response);
            }
        }
    }
}
```

The trick is here:
```
if(!Chat::where('uuid', $this->uuid)->value('stream')) {
   break 2;
}
```

So make sure every chat initiation creates new data in the Chat model in your application.
